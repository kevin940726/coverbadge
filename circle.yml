machine:
  timezone:
    Asia/Taipei
  node:
    version: 6.10.3
  environment:
    PATH: "${PATH}:${HOME}/${CIRCLE_PROJECT_REPONAME}/node_modules/.bin"

general:
  artifacts:
    - "coverage"

dependencies:
  pre:
    - echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" >> ~/.npmrc
  override:
    - yarn
  cache_directories:
    - ~/.cache/yarn

test:
  override:
    - yarn test -- --ci --runInBand
    - yarn run coverbadge -- -s circle -u $CIRCLE_PROJECT_USERNAME -p $CIRCLE_PROJECT_REPONAME

deployment:
  npm:
    tag: /v[0-9]+(\.[0-9]+)*/
    commands:
      - npm publish
