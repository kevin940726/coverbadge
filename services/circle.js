const path = require('path');
const fs = require('fs');
const fetch = require('isomorphic-fetch');
const get = require('lodash.get');

const HOST = 'https://circleci.com/api/v1.1';

const api = endpoint => (
  fetch(`${HOST}/${endpoint}`, {
    headers: {
      'Content-Type': 'Application/json',
    },
  }).then(res => res.json())
);

const getBuildSummary = ({
  username,
  project,
  vcs = 'github',
  token = null,
}) => {
  const tokenQuery = token ? `circle-token=${token}&` : '';

  return api(`project/${vcs}/${username}/${project}?${tokenQuery}limit=1&filter=successful`)
};

const getBuildArtifacts = ({
  username,
  project,
  buildNum,
  vcs = 'github',
  token = null,
}) => {
  const tokenQuery = token ? `?circle-token=${token}` : '';

  return api(`project/${vcs}/${username}/${project}/${buildNum}/artifacts${tokenQuery}`);
};

getLastBadgeURL = (artifacts, outputPath) => {
  const artifact = artifacts.find(art => art.path.indexOf(outputPath) >= 0);
  
  return artifact.url;
};

const downloadBadge = (url, outputPath) => {
  return fetch(url)
    .then(res => res.text())
    .then((svg) => {
      fs.writeFileSync(path.resolve(process.cwd(), outputPath), svg);
    });
}

const circle = ({
  username,
  project,
  token = null,
  vcs = 'github',
  outputPath,
}) => {
  const payload = {
    username,
    project,
    token,
    vcs,
  };

  return getBuildSummary(payload)
    .then(summary => (
      getBuildArtifacts(Object.assign({}, payload, {
        buildNum: get(summary, [0, 'previous_successful_build', 'build_num']),
      }))
    ))
    .then(artifacts => getLastBadgeURL(artifacts, outputPath))
    .then(url => downloadBadge(url, outputPath))
};

module.exports = {
  circle,
};
