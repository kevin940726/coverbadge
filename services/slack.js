const formatMessage = (lastCoverage, coverage) => {
  const coverageText = `${coverage}%`;
  const distance = `${coverage > lastCoverage ? '+' : ''}${(coverage - lastCoverage).toFixed(2)}%`;

  let emoji = coverage === 100 ? ':100:' : ':bookmark:';
  let text = `Coverage remained the same at *${coverage}%*.`;

  if (lastCoverage > coverage) {
    emoji = ':warning:';
    text = `Coverage decreased (*${distance}*) to *${coverageText}*.`;
  } else if (lastCoverage < coverage) {
    emoji = coverage === 100 ? ':100:' : ':tada:';
    text = `Coverage increased (*${distance}*) to *${coverageText}*.`;
  }

  return {
    emoji,
    text,
  };
};

const sendSlackWebhook = (webhook, lastCoverage, coverage) => {
  const { emoji, text } = formatMessage(lastCoverage, coverage);

  return fetch(webhook, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      username: 'coverbadge',
      icon_emoji: emoji,
      text,
    }),
  });
};

module.exports = {
  formatMessage,
  sendSlackWebhook,
};
