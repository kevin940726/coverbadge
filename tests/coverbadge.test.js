import cli from '../bin/coverbadge';
import { coverbadge } from '../';
import { circle } from '../services/circle';

jest.mock('../', () => ({
  coverbadge: jest.fn(),
}));

jest.mock('../services/circle', () => ({
  circle: jest.fn(() => Promise.resolve()),
}));

afterEach(() => {
  coverbadge.mockClear();
  circle.mockClear();
});

afterAll(() => {
  process.stdin.destroy();
});

describe('cli', () => {
  it('should pass default object if no options', async () => {
    const lcov = 'lcov';

    await cli(lcov);

    expect(coverbadge).not.toHaveBeenCalled();
  });

  it('should not do anything if no outputPath', async () => {
    const lcov = 'lcov';
    const options = {
      o: '',
    };

    await cli(lcov, options);

    expect(coverbadge).not.toHaveBeenCalled();
  });
});

describe('coverbadge', () => {
  it('should call coverbadge when there is service', async () => {
    const lcov = 'lcov';
    const outputPath = './output/path';
    const options = {
      o: outputPath, 
      s: circle,
    };

    await cli(lcov, options);

    expect(coverbadge).toHaveBeenCalledWith(lcov, outputPath);
  });

  it('should still call coverbadge when no service', async () => {
    const lcov = 'lcov';
    const outputPath = './output/path';
    const options = {
      o: outputPath, 
    };

    await cli(lcov, options);

    expect(coverbadge).toHaveBeenCalledWith(lcov, outputPath);
  });

  it('should still call coverbadge when there is service and the service fail', async () => {
    circle.mockImplementationOnce(() => Promise.reject());

    const lcov = 'lcov';
    const outputPath = './output/path';
    const options = {
      o: outputPath, 
      s: 'circle',
    };

    await cli(lcov, options);

    expect(coverbadge).toHaveBeenCalledWith(lcov, outputPath);
  });
});

describe('circle', () => {
  it('should call circle with options', async () => {
    const lcov = 'lcov';
    const outputPath = './output/path';
    const username = 'username';
    const project = 'project';
    const vcs = 'github';

    const options = {
      o: outputPath, 
      s: 'circle',
      u: username,
      p: project,
      vcs,
    };

    await cli(lcov, options);

    expect(circle).toHaveBeenCalledWith({
      username,
      project,
      outputPath,
      vcs,
      token: null,
    });
  });

  it('should call circle with token', async () => {
    const lcov = 'lcov';
    const outputPath = './output/path';
    const username = 'username';
    const project = 'project';
    const vcs = 'github';
    const token = 'token';

    const options = {
      o: outputPath, 
      s: 'circle',
      u: username,
      p: project,
      t: token,
      vcs,
    };

    await cli(lcov, options);

    expect(circle).toHaveBeenCalledWith({
      username,
      project,
      outputPath,
      vcs,
      token,
    });
  });
});
