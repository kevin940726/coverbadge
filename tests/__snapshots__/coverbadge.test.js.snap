// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displayCoverageInfo should display correct info when decreased 1`] = `"⚠️  [31mCoverage decreased ([1m-11.00%[22m) to [4m[1m55%[22m[24m.[39m"`;

exports[`displayCoverageInfo should display correct info when increased 1`] = `"🎉  [32mCoverage increased ([1m+11.00%[22m) to [4m[1m66%[22m[24m.[39m"`;

exports[`displayCoverageInfo should display correct info when increased and hit 100% 1`] = `"💯  [32mCoverage increased ([1m+45.00%[22m) to [4m[1m100%[22m[24m.[39m"`;

exports[`displayCoverageInfo should display correct info when remained 1`] = `"🔖  [32mCoverage remained the same at [4m[1m55%[22m[24m.[39m"`;

exports[`displayCoverageInfo should display correct info when remained and hit 100% 1`] = `"💯  [32mCoverage remained the same at [4m[1m100%[22m[24m.[39m"`;
